libraries/hotstuff/qc_chain.cpp

```diff
@@ -0,0 +1,1224 @@
+                              ("id",_id)
+                              ("proposal_id", proposal.proposal_id)
+                              ("justify_1", p->justify.proposal_id)
+                              ("justify_2", proposal.justify.proposal_id));
+
+         }
+
+         return; //already aware of proposal, nothing to do
+      }
+
+#ifdef QC_CHAIN_SIMPLE_PROPOSAL_STORE
+      ps_height_iterator psh_it = _proposal_stores_by_height.find( proposal.get_height() );
+      if (psh_it != _proposal_stores_by_height.end())
+      {
+         proposal_store & pstore = psh_it->second;
+         ps_iterator ps_it = pstore.begin();
+         while (ps_it != pstore.end())
+         {
+            hs_proposal_message & existing_proposal = ps_it->second;
+#else
+      //height is not necessarily unique, so we iterate over all prior proposals at this height
+      auto hgt_itr = _proposal_store.get<by_proposal_height>().lower_bound( proposal.get_height() );
+      auto end_itr = _proposal_store.get<by_proposal_height>().upper_bound( proposal.get_height() );
+      while (hgt_itr != end_itr)
+      {
+         const hs_proposal_message & existing_proposal = *hgt_itr;
+#endif
+
+         if (_errors) ilog(" *** ${id} received a different proposal at the same height (${block_num}, ${phase_counter})",
+                           ("id",_id)
+                           ("block_num", existing_proposal.block_num())
+                           ("phase_counter", existing_proposal.phase_counter));
+
+         if (_errors) ilog(" *** Proposal #1 : ${proposal_id_1} Proposal #2 : ${proposal_id_2}",
+                           ("proposal_id_1", existing_proposal.proposal_id)
+                           ("proposal_id_2", proposal.proposal_id));
+
+#ifdef QC_CHAIN_SIMPLE_PROPOSAL_STORE
+            ++ps_it;
+         }
+      }
+#else
+         hgt_itr++;
+      }
+#endif
+
+      if (_log) ilog(" === ${id} received new proposal : block_num ${block_num} phase ${phase_counter} : proposal_id ${proposal_id} : parent_id ${parent_id} justify ${justify}",
+                     ("id", _id)
+                     ("block_num", proposal.block_num())
+                     ("phase_counter", proposal.phase_counter)
+                     ("proposal_id", proposal.proposal_id)
+                     ("parent_id", proposal.parent_id)
+                     ("justify", proposal.justify.proposal_id));
+
+      bool success = insert_proposal( proposal );
+      EOS_ASSERT( success , chain_exception, "internal error: duplicate proposal insert attempt" ); // can't happen unless bad mutex somewhere; already checked for this
+
+      //if I am a finalizer for this proposal and the safenode predicate for a possible vote is true, sign
+      bool am_finalizer = am_i_finalizer();
+      bool node_safe = is_node_safe(proposal);
+      bool signature_required = am_finalizer && node_safe;
```

Comment:
This would be more efficient as `bool signature_required = am_i_finalizer() && is_node_safe(proposal);` as the `is_node_safe` will not be called unless `am_i_finalizer()` is true.